// @cypress/webpack-dev-server/dist/makeDefaultWebpackConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeDefaultWebpackConfig = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const debug_1 = tslib_1.__importDefault(require("debug"));
const debug = (0, debug_1.default)('cypress:webpack-dev-server:makeDefaultWebpackConfig');
const OUTPUT_PATH = path_1.default.join(__dirname, 'dist');
/**
 * @returns {import('webpack').Configuration}
 * @internal
 */
function makeDefaultWebpackConfig(config) {
    const { module: _HtmlWebpackPlugin, packageJson: { version }, importPath, } = config.sourceWebpackModulesResult.htmlWebpackPlugin;
    const indexHtmlFile = config.devServerConfig.cypressConfig.indexHtmlFile;
    const isRunMode = config.devServerConfig.cypressConfig.isTextTerminal;
    const HtmlWebpackPlugin = _HtmlWebpackPlugin;
    debug(`Using HtmlWebpackPlugin version ${version} from ${importPath}`);
    const optimization = {};
    if (config.sourceWebpackModulesResult.webpack.majorVersion === 5) {
        optimization.emitOnErrors = true;
    }
    else {
        optimization.noEmitOnErrors = false;
    }
    // To prevent files from being tree shaken by webpack, we set optimization.sideEffects: false ensuring that
    // webpack does not recognize the sideEffects flag in the package.json and thus files are not unintentionally
    // dropped during testing in production mode.
    optimization.sideEffects = false;
    const finalConfig = {
        mode: 'development',
        optimization: Object.assign(Object.assign({}, optimization), { splitChunks: {
                chunks: 'all',
            } }),
        output: {
            filename: '[name].js',
            path: OUTPUT_PATH,
        },
        plugins: [
            new HtmlWebpackPlugin(Object.assign({ template: indexHtmlFile }, (config.devServerConfig.framework === 'angular' ? { scriptLoading: 'module', base: '/__cypress/src/' } : {}))),
        ],
        devtool: 'inline-source-map',
    };
    if (isRunMode) {
        // Disable file watching when executing tests in `run` mode
        finalConfig.watchOptions = {
            ignored: '**/*',
        };
    }
    if (config.sourceWebpackModulesResult.webpackDevServer.majorVersion === 4) {
        return Object.assign(Object.assign({}, finalConfig), { devServer: {
                client: {
                    overlay: false,
                },
            } });
    }
    // @ts-ignore
    return Object.assign(Object.assign({}, finalConfig), { devServer: {
            overlay: false,
        } });
}
exports.makeDefaultWebpackConfig = makeDefaultWebpackConfig;
